### PR TITLE
add empty tiles, raster / grid traits

### PR DESCRIPTION
**Pull request by jdroenner**

### CODE REVIEW

**Comment by jdroenner** [comment]
[no data tiles](https://trello.com/c/4JGiIhJO/173-no-data-tiles)

File: datatypes/src/raster/mod.rs [hunk 1]
```diff
@@ -15,45 +15,54 @@ pub use self::grid_bounds::{
...
-    /// returns the grid dimension object of type D: `GridDimension`
-    fn dimension(&self) -> D;
-    /// returns the optional  no-data value used for the raster
-    fn no_data_value(&self) -> Option<T>;
-    // Checks whether the value is no data
-    fn is_no_data(&self, value: T) -> bool {
-        self.no_data_value()
-            .map_or(false, |no_data_value| value == no_data_value)
-    }
+pub trait Raster<D: GridSize, T: Pixel, C>:
```
Comment (ChristianBeilschmidt) [inline datatypes/src/raster/mod.rs:54]: Maybe we should have the container as an associated type? :thinking: What do you think?

File: datatypes/src/raster/no_data_grid.rs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    /// Converts the data type of the raster by converting it pixel-wise
+    pub fn convert_dtype<To>(self) -> NoDataGrid<D, To>
+    where
+        T: AsPrimitive<To> + Copy + 'static,
+        To: Copy + 'static,
+    {
+        NoDataGrid::new(self.shape, self.no_data_value.as_())
+            .expect("grid array type conversion cannot fail")
```
Comment (ChristianBeilschmidt) [inline datatypes/src/raster/no_data_grid.rs:53]: Why don't you just construct the type without using the constructor?

File: datatypes/src/raster/no_data_grid.rs [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    /// # Errors
+    ///
+    /// This constructor fails if the data container's capacity is different from the grid's dimension number
+    ///
+    pub fn new(shape: D, no_data_value: T) -> Result<Self> {
+        Ok(Self {
+            shape,
+            no_data_value,
+        })
+    }
```
Comment (ChristianBeilschmidt) [inline datatypes/src/raster/no_data_grid.rs:44]: Why does it return a result? It cannot fail.

File: datatypes/src/raster/operations/blit.rs [hunk 4]
```diff
@@ -37,23 +36,13 @@ impl<T: Pixel> Blit<RasterTile2D<T>> for RasterTile2D<T> {
...
-            data: source.grid_array.data,
-        };
+        let shifted_source = source.grid_array.shift_by_offset(offset);
 
         self.grid_array.grid_blit_from(shifted_source)?;
         Ok(())
     }
 }
-
+/*/
```
Comment (ChristianBeilschmidt) [inline datatypes/src/raster/operations/blit.rs:45]: Why did you remove it kind of but not completely? Don't we need the blit operation?

File: datatypes/src/raster/operations/grid_blit.rs [hunk 5]
```diff
@@ -70,14 +68,55 @@ where
...
+                for x in overlap_x_start..overlap_x_start + overlap_x_size as isize {
+                    self.set_at_grid_index_unchecked(
+                        [y, x],
+                        other.get_at_grid_index_unchecked([y, x]),
+                    )
+                }
+            }
+        }
+        Ok(())
+    }
```
Comment (ChristianBeilschmidt) [inline datatypes/src/raster/operations/grid_blit.rs:97]: It seems this method cannot fail, right?

File: datatypes/src/raster/operations/grid_blit.rs [hunk 6]
```diff
@@ -70,14 +68,55 @@ where
...
+{
+    fn grid_blit_from(&mut self, other: GridOrEmpty<D, T>) -> Result<()> {
+        match other {
+            GridOrEmpty::Grid(g) => self.grid_blit_from(g),
+            GridOrEmpty::Empty(n) => self.grid_blit_from(n),
+        }
+    }
+}
+
+impl<T> GridBlit<Grid3D<T>, T> for Grid3D<T>
```
Comment (ChristianBeilschmidt) [inline datatypes/src/raster/operations/grid_blit.rs:115]: I guess you should test the 3D blit like you did for the 2D case.

File: datatypes/src/raster/raster_tile.rs [hunk 7]
```diff
@@ -11,32 +12,70 @@ use crate::util::Result;
...
+
 /// A `RasterTile2D` is the main type used to iterate over tiles of 2D raster data
 #[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
 #[serde(rename_all = "camelCase")]
-pub struct RasterTile<D, T>
-where
-    D: GridSize + GridSpaceToLinearSpace,
-    T: Pixel,
-{
+pub struct BaseTile<G> {
```
Comment (ChristianBeilschmidt) [inline datatypes/src/raster/raster_tile.rs:26]: Why is it a `BaseTile`? You also did not change the doc-comment.

File: datatypes/src/raster/raster_tile.rs [hunk 8]
```diff
@@ -11,32 +12,70 @@ use crate::util::Result;
...
     pub time: TimeInterval,
     /// The tile position is the position of the tile in the gird of tiles with origin at the origin of the global_geo_transform.
     /// This is NOT a pixel position inside the tile.
     pub tile_position: GridIdx2D,
     /// The global geotransform to transform pixels into geographic coordinates
     pub global_geo_transform: GeoTransform,
-    /// The data of the `RasterTile` is stored as `Grid`
-    pub grid_array: Grid<D, T>,
+    /// The data of the `RasterTile` is stored as `Grid` or `NoDataGrid`. `GridOrEmpty` allows a combination of both.
+    pub grid_array: G,
```
Comment (ChristianBeilschmidt) [inline datatypes/src/raster/raster_tile.rs:35]: Why not just use that enum?

File: operators/src/plot/histogram.rs [hunk 9]
```diff
@@ -264,10 +267,12 @@ impl HistogramRasterQueryProcessor {
...
             while let Some(tile) = input.next().await {
-                let tile = tile?;
-
-                computed_metadata
-                    .add_raster_batch(&tile.grid_array.data, tile.grid_array.no_data_value);
+                match tile?.grid_array {
+                    geoengine_datatypes::raster::GridOrEmpty::Grid(g) => {
+                        computed_metadata.add_raster_batch(&g.data, g.no_data_value);
+                    }
+                    geoengine_datatypes::raster::GridOrEmpty::Empty(_) => {} // TODO: find out if we really do nothing for empty tiles?
```
Comment (ChristianBeilschmidt) [inline operators/src/plot/histogram.rs:274]: I guess we need to increase no data value counts and you already did implement a method for that.

File: datatypes/src/raster/operations/blit.rs (hunk 4 above)
Comment (jdroenner) [reply datatypes/src/raster/operations/blit.rs:45]: i wanted to discuss this. we do not use the code anywhere and it will require us to first materialize each tile since blitting into is only possible on materialized tiles. I can add this or remove the code.

File: datatypes/src/raster/operations/grid_blit.rs (hunk 5 above)
Comment (jdroenner) [reply datatypes/src/raster/operations/grid_blit.rs:97]: we use the unchecked get / set methods since we checked the bounds earlier. So i guess it should not fail

File: datatypes/src/raster/grid_traits.rs [hunk 10]
```diff
@@ -126,3 +128,50 @@ pub trait GridIndexAccess<T, I> {
...
+    fn grid_shape_array(&self) -> Self::ShapeArray;
+
+    fn grid_shape(&self) -> GridShape<Self::ShapeArray> {
+        GridShape::new(self.grid_shape_array())
+    }
+}
+
+/// Provides the the value used to represent a no data entry.
+pub trait NoDataValue {
+    type NoDataType: PartialEq;
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply datatypes/src/raster/grid_traits.rs:145]: it is not needed to be pixel here. all it needs to be is copy

File: datatypes/src/raster/raster_tile.rs (hunk 8 above)
Comment (jdroenner) [reply datatypes/src/raster/raster_tile.rs:35]: as explained above, this allows to use the BaseTile for both, the NoData and the Materialized tiles.

File: datatypes/src/raster/no_data_grid.rs [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+}
+
+pub type NoDataGrid1D<T> = NoDataGrid<GridShape1D, T>;
+pub type NoDataGrid2D<T> = NoDataGrid<GridShape2D, T>;
+pub type NoDataGrid3D<T> = NoDataGrid<GridShape3D, T>;
+
+impl<D, T> NoDataGrid<D, T>
+where
+    D: GridSize,
+    T: Clone,
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply datatypes/src/raster/no_data_grid.rs:31]: i think it would be nice to have a Grid of Coordinates. So no, i do not think it should be pixel. While i don't see a reason to have a NoDataGRid of something other then pixels it should be the same as for Grid.

File: operators/src/adapters/raster_overlap_adapter.rs [hunk 12]
```diff
@@ -295,18 +304,23 @@ where
...
+    match insert_projected_pixels(&mut materialized_accu_tile, &tile, accu_companion.iter()) {
+        Ok(_) => Ok((materialized_accu_tile.into(), accu_companion)),
         Err(error) => Err(error),
     }
 }
 
 /// This method takes two tiles and a map from `GridIdx2D` to `Coordinate2D`. Then for all `GridIdx2D` we set the values from the corresponding coordinate in the source tile.
 pub fn insert_projected_pixels<'a, T: Pixel, I: Iterator<Item = &'a (GridIdx2D, Coordinate2D)>>(
-    target: &mut RasterTile2D<T>,
+    target: &mut MaterializedRasterTile2D<T>,
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply operators/src/adapters/raster_overlap_adapter.rs:323]: yes, keep in mind that the target here is the accu tile. In line 309 we skip the whole insertion of anything if the input tile is empty. this allows to keep the target to hold a NoDataGRid as long as possible. If the input is not empty then we need to write individual pixels. Therefore, in line 313 we transform the accu tile into a MAterializedTile (no work if it is already). Then we call this method with a writable tile as target (accu) and any tile as source (can read from NoDataTile).

File: datatypes/src/raster/grid_or_empty.rs [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+{
+    type IndexArray = I;
+
+    fn min_index(&self) -> GridIdx<Self::IndexArray> {
+        todo!()
+    }
+
+    fn max_index(&self) -> GridIdx<Self::IndexArray> {
+        todo!()
+    }
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply datatypes/src/raster/grid_or_empty.rs]: 8e9f8f5

File: datatypes/src/operations/image/to_png.rs [hunk 14]
```diff
@@ -68,7 +68,8 @@ fn create_rgba_image<P: Pixel + RgbaTransmutable, N: Fn(P) -> bool>(
 
 impl<T: Pixel> ToPng for RasterTile2D<T> {
     fn to_png(&self, width: u32, height: u32, colorizer: &Colorizer) -> Result<Vec<u8>> {
-        self.grid_array.to_png(width, height, colorizer)
+        let temp_copy = self.clone().into_materialized_tile(); //FIXME: avoid the clone!
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply datatypes/src/operations/image/to_png.rs]: e69655e

File: datatypes/src/raster/mod.rs (hunk 1 above)
Comment (jdroenner) [reply datatypes/src/raster/mod.rs]: very good idea

68824c5

File: datatypes/src/raster/no_data_grid.rs (hunk 2 above)
Comment (jdroenner) [reply datatypes/src/raster/no_data_grid.rs]: no longer a problem since it cant fail

File: datatypes/src/raster/no_data_grid.rs (hunk 3 above)
Comment (jdroenner) [reply datatypes/src/raster/no_data_grid.rs]: 4516158

File: operators/src/plot/histogram.rs [hunk 15]
```diff
@@ -303,9 +308,12 @@ impl HistogramRasterQueryProcessor {
             let mut query = processor.query(query, ctx)?;
 
             while let Some(tile) = query.next().await {
-                let tile = tile?;
 
-                histogram.add_raster_data(&tile.grid_array.data, tile.grid_array.no_data_value);
+
+                match tile?.grid_array {
+                    geoengine_datatypes::raster::GridOrEmpty::Grid(g) => histogram.add_raster_data(&g.data, g.no_data_value),
+                    geoengine_datatypes::raster::GridOrEmpty::Empty(n) => histogram.add_nodata_batch(n.number_of_elements() as u64) // TODO: why u64?
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply operators/src/plot/histogram.rs:315]: but why not usize?

File: operators/src/mock/mock_raster_source.rs [hunk 16]
```diff
@@ -166,11 +166,13 @@ mod tests {
                         "yPixelSize": -1.0
                     },
                     "gridArray": {
-                        "shape": {
-                            "shapeArray": [3, 2]
-                        },
-                        "data": [1, 2, 3, 4, 5, 6],
-                        "noDataValue": null
+                        "grid": {
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply operators/src/mock/mock_raster_source.rs]: e467877

File: operators/src/plot/histogram.rs (hunk 9 above)
Comment (jdroenner) [reply operators/src/plot/histogram.rs:274]: the add_raster_batch does this:
```
                if v == no_data {
                    continue;
                }
```

File: operators/src/plot/temporal_raster_mean_plot.rs [hunk 17]
```diff
@@ -131,6 +131,12 @@ impl<P: Pixel> MeanRasterPixelValuesOverTimeQueryProcessor<P> {
         while let Some(tile) = tile_stream.next().await {
             let tile = tile?;
 
+            if tile.grid_array.is_empty() {
+                continue;
+            }
+
+            let tile = tile.into_materialized_tile(); // TODO: is this required?
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply operators/src/plot/temporal_raster_mean_plot.rs:138]: i think we should remove the todo. we check if the tile is empty and skip in the lines above.

File: datatypes/src/raster/grid.rs [hunk 18]
```diff
@@ -388,6 +391,53 @@ where
...
+    type ShapeArray = D::ShapeArray;
+
+    fn grid_shape_array(&self) -> Self::ShapeArray {
+        self.shape.axis_size()
+    }
+}
+
+impl<D, T> NoDataValue for Grid<D, T>
+where
+    T: PartialEq + Clone,
```
[the remark this answers is not in the document]
Comment (ChristianBeilschmidt) [reply datatypes/src/raster/grid.rs:411]: I expect NoDataValues to be Copy-Types so I would not like arbitrary Heap-Types to be NoDataValues.

File: datatypes/src/operations/image/to_png.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (ChristianBeilschmidt) [reply datatypes/src/operations/image/to_png.rs]: Looks correct, could you create a test that checks that there is an empty png output?

File: datatypes/src/raster/grid_traits.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (ChristianBeilschmidt) [reply datatypes/src/raster/grid_traits.rs:145]: I expect NoDataValues to be Copy-Types so I would not like arbitrary Heap-Types to be NoDataValues.

File: datatypes/src/raster/no_data_grid.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (ChristianBeilschmidt) [reply datatypes/src/raster/no_data_grid.rs:31]: Maybe then also Copy here, since Coordinates are Copy.

File: operators/src/plot/histogram.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (ChristianBeilschmidt) [reply operators/src/plot/histogram.rs:315]: Good point, I don't know, we could change it I guess :shrug: .

[the remark this answers is not in the document]
Comment (jdroenner) [reply operators/src/plot/histogram.rs:315]: 24f5725

File: datatypes/src/raster/grid_traits.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (jdroenner) [reply datatypes/src/raster/grid_traits.rs]: 24f5725

File: datatypes/src/raster/no_data_grid.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (jdroenner) [reply datatypes/src/raster/no_data_grid.rs]: 24f5725

File: datatypes/src/operations/image/to_png.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (jdroenner) [reply datatypes/src/operations/image/to_png.rs]: 9b40d98

File: datatypes/src/raster/empty_grid.rs [hunk 19]
```diff
@@ -16,19 +16,19 @@ use snafu::ensure;
...
     pub shape: D,
     pub no_data_value: T,
 }
 
-pub type NoDataGrid1D<T> = NoDataGrid<GridShape1D, T>;
-pub type NoDataGrid2D<T> = NoDataGrid<GridShape2D, T>;
-pub type NoDataGrid3D<T> = NoDataGrid<GridShape3D, T>;
+pub type NoDataGrid1D<T> = EmptyGrid<GridShape1D, T>;
+pub type NoDataGrid2D<T> = EmptyGrid<GridShape2D, T>;
+pub type NoDataGrid3D<T> = EmptyGrid<GridShape3D, T>;
```
Comment (ChristianBeilschmidt) [inline datatypes/src/raster/empty_grid.rs:26]: either `EmptyGrid` and `EmptyGrid1D' (etc.) or `NoDataGrid1D` and `NoDataGrid`

**Review by ChristianBeilschmidt — CHANGES_REQUESTED** [review]
just one thing is left.

File: datatypes/src/raster/empty_grid.rs (hunk 19 above)
Comment (jdroenner) [reply datatypes/src/raster/empty_grid.rs]: 03e07cc

**Review by ChristianBeilschmidt — APPROVED** [review]
